### PR TITLE
Preview of Binder IPC driver in Rust.

### PR DESCRIPTION
**Pull request by wedsonaf**

### CODE REVIEW

File: drivers/android/allocation.rs [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+                self.process.update_ref(handle, false, strong)
+            }
+            _ => Ok(()),
+        }
+    }
+}
+
+impl Drop for Allocation<'_> {
+    fn drop(&mut self) {
+        if self.free_on_drop {
```
Comment (ojeda) [inline drivers/android/allocation.rs]: A level of indentation can be removed by negating this one.

File: drivers/android/defs.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use core::ops::{Deref, DerefMut};
+use kernel::{
+    bindings,
+    user_ptr::{ReadableFromBytes, WritableToBytes},
+};
+
+pub const BR_ERROR: u32 = bindings::binder_driver_return_protocol_BR_ERROR;
+pub const BR_OK: u32 = bindings::binder_driver_return_protocol_BR_OK;
```
Comment (ojeda) [inline drivers/android/defs.rs]: Perhaps a macro to generate all of these easily? i.e.

    generate_br_constants!(TRANSACTION, REPLY, DEAD_REPLY ......)

File: drivers/android/rust_binder.rs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+}
+
+enum Either<L, R> {
+    Left(L),
+    Right(R),
+}
+
+trait DeliverToRead {
+    /// Performs work. Returns true if remaining work items in the queue should be processed
+    /// immediately, or false if it should return to caller before processing additional work
```
Comment (ojeda) [inline drivers/android/rust_binder.rs:45]:
```suggestion
    /// Performs work. Returns `true` if remaining work items in the queue should be processed
    /// immediately, or `false` if it should return to caller before processing additional work
```

File: rust/kernel/file_operations.rs [hunk 4]
```diff
@@ -506,6 +588,22 @@ pub trait FileOperations: Send + Sync + Sized {
     fn fsync(&self, _file: &File, _start: u64, _end: u64, _datasync: bool) -> KernelResult<u32> {
         Err(Error::EINVAL)
     }
+
+    /// Maps areas of the caller's virtual memory with device/file memory.
+    ///
+    /// Corresponds to the `mmap` function pointer in `struct file_operations`.
+    /// TODO: wrap vm_area_struct so that we don't have to expose it.
```
Comment (ojeda) [inline rust/kernel/file_operations.rs]:
```suggestion
    /// TODO: wrap `vm_area_struct` so that we don't have to expose it.
```

File: drivers/android/node.rs [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+        }
+    }
+
+    pub(crate) fn init(self: Pin<&Self>) {
+        // SAFETY: `inner` is pinned when `self` is.
+        let inner = unsafe { self.map_unchecked(|s| &s.inner) };
+        kernel::spinlock_init!(inner, "NodeDeath::inner");
+    }
+
+    /// Sets the cleared flag to true.
```
Comment (ojeda) [inline drivers/android/node.rs]: There are a bunch of these, but it is a very minor doc nit, so feel free to ignore for the moment.
```suggestion
    /// Sets the cleared flag to `true`.
```

File: drivers/android/range_alloc.rs [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    pub(crate) fn reservation_commit(&mut self, offset: usize, data: Option<T>) -> KernelResult {
+        // TODO: This is currently O(n), make it O(1).
+        let mut cursor = self.find_at_offset(offset).ok_or(Error::ENOENT)?;
+        let desc = cursor.current().unwrap();
+        desc.state = DescriptorState::Allocated;
+        desc.data = data;
+        Ok(())
+    }
+
+    /// Takes an entry at the given offset from the Allocated state to Reserved.
```
Comment (ojeda) [inline drivers/android/range_alloc.rs]:
```suggestion
    /// Takes an entry at the given offset from [`DescriptorState::Allocated`]
    /// to [`DescriptorState::Reserved`].
```

File: drivers/android/raw_list.rs [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+    }
+
+    fn remove_internal(&mut self, data: &G::EntryType) -> bool {
+        let links = unsafe { &mut *G::get_links(data).0.get() };
+        let next = if let Some(next) = links.next {
+            next
+        } else {
+            // Nothing to do if the entry is not on the list.
+            return false;
+        };
```
Comment (ojeda) [inline drivers/android/raw_list.rs:119]: Hopefully the `unless`-like proposal ends up being accepted in Rust so that we can do this pattern easily :-)

(Nothing to do here)

File: rust/kernel/file_operations.rs [hunk 8]
```diff
@@ -49,6 +49,50 @@ impl File {
...
+    ///
+    /// The pointer `ptr` must be either null or a valid pointer for the lifetime of the object.
+    unsafe fn from_ptr(ptr: *mut bindings::poll_table_struct) -> Self {
+        Self { ptr }
+    }
+
+    /// Associates the given file and condition variable to this poll table. It means notifying the
+    /// condition variable will notify the poll table as well; additionally, the association
+    /// between the condition variable and the file will automatically be undone by the kernel when
+    /// the file is destructed. To unilaretally remove the association before then, one can call
```
Comment (ojeda) [inline rust/kernel/file_operations.rs]:
```suggestion
    /// the file is destructed. To unilaterally remove the association before then, one can call
```

File: rust/kernel/file_operations.rs [hunk 9]
```diff
@@ -291,6 +365,12 @@ pub struct ToUse {
 
     /// The `fsync` field of [`struct file_operations`].
     pub fsync: bool,
+
+    /// The `mmap` field of `[struct file_operations]`.
```
Comment (ojeda) [inline rust/kernel/file_operations.rs]:
```suggestion
    /// The `mmap` field of [`struct file_operations`].
```

File: rust/kernel/file_operations.rs [hunk 10]
```diff
@@ -291,6 +365,12 @@ pub struct ToUse {
...
+    pub mmap: bool,
+
+    /// The `poll` field of `[struct file_operations]`.
```
Comment (ojeda) [inline rust/kernel/file_operations.rs]:
```suggestion
    /// The `poll` field of [`struct file_operations`].
```

**Review by ojeda — COMMENTED** [review]
This is great!

A quick few comments. In general, it looks fine for the RFC.

File: drivers/android/defs.rs (hunk 2 above)
Comment (wedsonaf) [reply drivers/android/defs.rs]: The problem is that given a name, say, `BR_TRANSACTION`, we need to generate a new identifier `bindings::binder_driver_return_protocol_BR_TRANSACTION`, which is a concatenation. It looks like we could use `concat_idents!` but it's unstable.

I'd rather leave this as is for now and focus on the other missing features.

File: drivers/android/allocation.rs (hunk 1 above)
Comment (wedsonaf) [reply drivers/android/allocation.rs]: Done.

File: rust/kernel/file_operations.rs (hunk 4 above)
Comment (wedsonaf) [reply rust/kernel/file_operations.rs]: Done.

File: drivers/android/node.rs (hunk 5 above)
Comment (wedsonaf) [reply drivers/android/node.rs]: Done.

File: drivers/android/range_alloc.rs (hunk 6 above)
Comment (wedsonaf) [reply drivers/android/range_alloc.rs]: Done.

File: rust/kernel/file_operations.rs (hunk 8 above)
Comment (wedsonaf) [reply rust/kernel/file_operations.rs]: Done.

File: rust/kernel/file_operations.rs (hunk 9 above)
Comment (wedsonaf) [reply rust/kernel/file_operations.rs]: Done.

File: rust/kernel/file_operations.rs (hunk 10 above)
Comment (wedsonaf) [reply rust/kernel/file_operations.rs]: Done.

File: drivers/android/defs.rs (hunk 2 above)
Comment (ojeda) [reply drivers/android/defs.rs]: Of course, don't worry about it. I was thinking more from the PoV of readers of the RFC, thinking "Rust cannot do this with a macro?!?!" (because the "block" of definitions stands out when scrolling).

Comment (wedsonaf) [reply drivers/android/defs.rs]: I see. I understand the concern now. Let me see if I can do it with a macro.

Comment (wedsonaf) [reply drivers/android/defs.rs]: Unfortunately, it looks like we need proc macros to do this, from https://doc.rust-lang.org/std/macro.concat_idents.html:
> That means while you may use this macro for referring to existing variables, functions or modules etc, you cannot define a new one with it.

We can either write one or import some crate that does this for us. Both options are undesirable.

Do you see any other way to improve this?

Comment (ojeda) [reply drivers/android/defs.rs]: Bummer... Nah, I wouldn't spend more time on this. Perhaps just add `TODO` comment so that people now when reading.

Comment (ojeda) [reply drivers/android/defs.rs]: We could always run the C preprocessor on Rust files, though. 😈

Comment (wedsonaf) [reply drivers/android/defs.rs]: We actually only _refer_ to concatenated names, it may be possible to use macros. Let me try again.

Comment (ojeda) [reply drivers/android/defs.rs]: If you put the `BR_` in the arguments, yeah, it would only be a reference, indeed. Good catch.

Comment (wedsonaf) [reply drivers/android/defs.rs]: We needed to include `BR_` anyway because we also have `BC_`. Also needed to import `bindings::*` but it works!

Comment (ojeda) [reply drivers/android/defs.rs]: Looks nice :)

File: drivers/android/allocation.rs [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+    fn drop(&mut self) {
+        if !self.free_on_drop {
+            return;
+        }
+
+        if let Some(info) = &self.allocation_info {
+            let view = AllocationView::new(self, info.offsets.start);
+            for i in info.offsets.clone().step_by(size_of::<usize>()) {
+                if self.cleanup_object(i, &view).is_err() {
+                    println!("Error cleaning up object at offset {}", i)
```
Comment (ojeda) [inline drivers/android/allocation.rs]: `err!` now (or `warn!` or `crit!` depending how you see it).

File: drivers/android/thread.rs [hunk 12]
```diff
@@ -0,0 +1,821 @@
...
+            }
+            bindings::BINDER_TYPE_WEAK_HANDLE | bindings::BINDER_TYPE_HANDLE => {
+                let strong = header.type_ == bindings::BINDER_TYPE_HANDLE;
+                view.transfer_binder_object(offset, strong, |obj| {
+                    // SAFETY: The type is `BINDER_TYPE_{WEAK_}HANDLE`, so `handle` is populated.
+                    let handle = unsafe { obj.__bindgen_anon_1.handle } as _;
+                    self.process.get_node_from_handle(handle, strong)
+                })?;
+            }
+            _ => println!("Unsupported binder object type: {:x}", header.type_),
```
Comment (ojeda) [inline drivers/android/thread.rs]: `warn!`?

**Review by ojeda — COMMENTED** [review]
Since #120 is now in, you will need to change the `println!`s, which the CI didn't catch because it is not enabled in the configs :)

File: drivers/android/allocation.rs (hunk 11 above)
Comment (wedsonaf) [reply drivers/android/allocation.rs]: Done.

File: drivers/android/thread.rs (hunk 12 above)
Comment (wedsonaf) [reply drivers/android/thread.rs]: Done.

**Comment by wedsonaf** [comment]
I squashed all the commits into one and rebased.

File: drivers/android/thread.rs [hunk 13]
```diff
@@ -0,0 +1,821 @@
...
+        unsafe { Pin::new_unchecked(&self.work_condvar) }
+    }
+
+    pub(crate) fn set_current_transaction(&self, transaction: Arc<Transaction>) {
+        self.inner.lock().current_transaction = Some(transaction);
+    }
+
+    /// Attemps to fetch a work item from the thread-local queue. The behaviour if it the queue is
+    /// empty depends on `wait`: if it is true, the function waits for some work to be queued (or a
+    /// signal); otherwise it returns indicating that none is avaible.
```
Comment (ojeda) [inline drivers/android/thread.rs]:
```suggestion
    /// Attempts to fetch a work item from the thread-local queue. The behaviour if the queue is
    /// empty depends on `wait`: if it is true, the function waits for some work to be queued (or a
    /// signal); otherwise it returns indicating that none is available.
```

File: drivers/android/thread.rs [hunk 14]
```diff
@@ -0,0 +1,821 @@
...
+            let signal_pending = cv.wait(&mut inner);
+            inner.looper_flags &= !LOOPER_WAITING;
+
+            if signal_pending {
+                return Err(Error::ERESTARTSYS);
+            }
+        }
+    }
+
+    /// Attemps to fetch a work item from the thread-local queue, falling back to the process-wide
```
Comment (ojeda) [inline drivers/android/thread.rs]:
```suggestion
    /// Attempts to fetch a work item from the thread-local queue, falling back to the process-wide
```

File: drivers/android/thread.rs [hunk 15]
```diff
@@ -0,0 +1,821 @@
...
+            if inner.is_dead {
+                return Err(BinderError::new_dead());
+            }
+            inner.push_work(work);
+        }
+        self.pinned_condvar().notify_one();
+        Ok(())
+    }
+
+    /// Attemps to push to given work item to the thread if it's a looper thread (i.e., if it's
```
Comment (ojeda) [inline drivers/android/thread.rs]:
```suggestion
    /// Attempts to push to given work item to the thread if it's a looper thread (i.e., if it's
```

File: drivers/android/thread.rs [hunk 16]
```diff
@@ -0,0 +1,821 @@
...
+        (|| -> BinderResult<_> {
+            let completion = Arc::try_new(DeliverCode::new(BR_TRANSACTION_COMPLETE))?;
+            let process = orig.from.process.clone();
+            let reply = Arc::try_new(Transaction::new_reply(self, process, tr)?)?;
+            self.inner.lock().push_work(completion);
+            orig.from.deliver_reply(Either::Left(reply), &orig);
+            Ok(())
+        })()
+        .map_err(|mut err| {
+            // At this point we only return BR_TRANSACTION_COMPLETE to the caller, and we must let
```
Comment (ojeda) [inline drivers/android/thread.rs]:
```suggestion
            // At this point we only return `BR_TRANSACTION_COMPLETE` to the caller, and we must let
```

File: drivers/android/thread.rs [hunk 17]
```diff
@@ -0,0 +1,821 @@
...
+            let reply = Either::Right(BR_FAILED_REPLY);
+            orig.from.deliver_reply(reply, &orig);
+            err.reply = BR_TRANSACTION_COMPLETE;
+            err
+        })
+    }
+
+    /// Determines the current top of the transaction stack. It fails if the top is in another
+    /// thread (i.e., this thread belongs to a stack but it has called another thread). The top is
+    /// `None` if the thread is not currently participating in a transaction stack.
```
Comment (ojeda) [inline drivers/android/thread.rs]:
```suggestion
    /// [`None`] if the thread is not currently participating in a transaction stack.
```

File: drivers/android/rust_binder.rs [hunk 18]
```diff
@@ -0,0 +1,216 @@
+// SPDX-License-Identifier: GPL-2.0
+
+#![no_std]
```
Comment (ojeda) [inline drivers/android/rust_binder.rs]: Perhaps a top-level description like:

    //! Binder -- the Android IPC mechanism.
    //!
    //! TODO: This module is a work in progress.

Comment (wedsonaf) [reply drivers/android/rust_binder.rs]: Done.

**Review by ojeda — APPROVED** [review]
LGTM -- I see you changed the `file_operations` bits we discussed the other day to `unsafe`, so it is fine.

I let @alex merge in case he wants to take a better look.

**Review by alex — COMMENTED** [review]
I only really reviewed the `rust/kernel/` part of this, but it looks ok to me. Can we please file a follow-up task to make the `mmap` API more idiomatic?

**Comment by wedsonaf** [comment]
> I only really reviewed the `rust/kernel/` part of this, but it looks ok to me. Can we please file a follow-up task to make the `mmap` API more idiomatic?

#178 for mmap and #179 for poll.

File: rust/kernel/bindings_helper.h [hunk 19]
```diff
@@ -10,6 +10,8 @@
 #include <linux/version.h>
 #include <linux/miscdevice.h>
 #include <linux/poll.h>
+#include <linux/mm.h>
+#include <uapi/linux/android/binder.h>
```
Comment (fbq) [inline rust/kernel/bindings_helper.h:14]: Sorry, late to party. But I think in the long term, maybe we don't want to put this `#include` in the `rust/kernel/bindings_helper.h`? Maybe we can support module-level `bindgen`, that is allowing a module to have its own `bindings_helper.h` and only generate when necessary.

Maybe open an issue?

Thoughts?

Comment (ojeda) [reply rust/kernel/bindings_helper.h:14]: +1, we discussed this the other day. There are a couple of bits of functionality that will need to be expanded to individual modules.

Comment (ojeda) [reply rust/kernel/bindings_helper.h:14]: Done: #181 for the documentation, #182 for the bindings.
